### PR TITLE
graph: squash commits

### PR DESCRIPTION
**Pull request by estib-vega**
Add an API that squashes commits using the new rebase engine.

### CODE REVIEW

File: crates/but-workspace/src/commit/squash_commits.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        MoveSubjectAboveTarget,
+        MoveSubjectBelowTarget,
+    }
+
+    fn determine_reorder_direction(
+        repo: &gix::Repository,
+        subject: &but_core::CommitOwned,
+        target: &but_core::CommitOwned,
+    ) -> Result<ReorderDirection> {
+        let merge_base = match repo.merge_base(subject.id, target.id) {
```
Comment (Byron) [inline crates/but-workspace/src/commit/squash_commits.rs:68]: Can you try to use `editor.workspace.graph.find_git_merge_base(a, b)` here?
It's faster and the graph is expected to have all knowledge it needs to make that computation.

File: crates/but-workspace/tests/workspace/commit/squash_commits.rs [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+    insta::assert_snapshot!(visualize_commit_graph_all(&repo)?, @"
+    * c9f444c (HEAD -> three) commit three
+    * 16fd221 (origin/two, two) commit two
+    * 8b426d0 (one) commit one
+    ");
+
+    let subject_id = repo.rev_parse_single("three")?.detach();
+    let target_id = repo.rev_parse_single("two")?.detach();
+    let subject_tree = repo.find_commit(subject_id)?.tree_id()?.detach();
```
Comment (Byron) [inline crates/but-workspace/tests/workspace/commit/squash_commits.rs:25]: The last time I tried I was able to remove the `detach()` calls.

**Review by Byron — APPROVED** [review]
I took a look at it merely from a graph + gix perspectives and left a few comments, with the big one to avoid `repo.merge_base()` if a workspace is present.

It's also good to see a couple of tests involving a workspace, showing that branches are simply left empty once their only commit is squashed, which seems desirable.

Also, when looking at how merge-base is used, maybe we want something like `graph.relation_between(a, b)` yielding `Disjoint, Ancestor, Descendant, Identity` or something, to make clear what's going on?

If you agree, maybe just implement this in `but-graph` - it's pre-approved :D.

File: crates/but-workspace/src/commit/squash_commits.rs (hunk 1 above)
Comment (estib-vega) [reply crates/but-workspace/src/commit/squash_commits.rs:68]: Will do, out of curiosity: Does the fact that the workspace is a projection not mean that the merge base provided here might not be the true merge base?

File: crates/but-workspace/tests/workspace/commit/squash_commits.rs (hunk 2 above)
Comment (estib-vega) [reply crates/but-workspace/tests/workspace/commit/squash_commits.rs:25]: Will try that

File: crates/but-workspace/src/commit/squash_commits.rs (hunk 1 above)
Comment (estib-vega) [reply crates/but-workspace/src/commit/squash_commits.rs:68]: Oh, `graph.find_git_merge_base` seems to be for segments, not commits like in this case

Comment (estib-vega) [reply crates/but-workspace/src/commit/squash_commits.rs:68]: I'm using it at the beginning, but if the commits are part of the same segment, then I'll still use the repo's merge base

Comment (Byron) [reply crates/but-workspace/src/commit/squash_commits.rs:68]: > Oh, `graph.find_git_merge_base` seems to be for segments, not commits like in this case

Yes, that's fine as every fork/split is happening in its own segments. Thus the segment you get always has a commit. You can see how it's already used in `but-graph`.

Comment (Byron) [reply crates/but-workspace/src/commit/squash_commits.rs:68]: > I'm using it at the beginning, but if the commits are part of the same segment, then I'll still use the repo's merge base

I think now I understand that you tried using it, but it didn't give the desired results. This is something to look into and would love to understand the problem - merge-base computation should just be the exact same, and for all I could tell, it is.
